text/0122-sdk-hub-scope-merge.md

```diff
@@ -0,0 +1,148 @@
+- Start Date: 2023-11-07
+- RFC Type: feature
+- RFC PR: https://github.com/getsentry/rfcs/pull/122
+- RFC Status: draft
+
+# Summary
+
+We want to streamline Hub & Scope usage for the SDK.
+
+# Motivation
+
+Generally, Hubs and Scopes are a very Sentry-specific concept, and can be a bit hard to explain. 
+Also, how Hub & Scope forking behaves is currently a bit tricky, and does not play well e.g. with how OpenTelemetry contexts work.
+
+This RFC aims to streamline this by merging the concepts of Hub and Scope into a singular Scope concept.
+
+# Background
+
+TODO
+
+# Proposed Solution
+
+Hubs will be removed as a concept for SDK users. 
+Instead, from a users perspective only Scopes remain, which will become the singular entity to hold context data etc.
+
+Scopes will be _similar_ to how they work today, but not entirely the same. 
+Scopes can have data (e.g. tags, user, ...) added to them the same way you can do today. 
+This RFC _does not_ aim to change any of the data that is kept on the scope and is applied to events.
+
+The following APIs will be removed:
+
+* `getCurrentHub()`
+* `configureScope()` (instead just get the scope and set on it directly)
+* Any APIs currently on the Hub only: 
+  * `hub.pushScope()`
+  * `hub.popScope()`
+  * `hub.isOlderThan()`
+  * `hub.bindClient()`
+  * `hub.getStack()`
+  * `hub.getStackTop()`
+  * `hub.run()` (use `withScope()` instead)
+
+Instead, we will introduce some new APIs:
+
+```ts
+// get the currently active scope. replacement for `getCurrentHub().getScope()`
+export function getScope(): Scope;
+
+// get the currently active client. May return a NOOP client. Replacement for `getCurrentHub().getClient()`.
+export function getClient(): Client;
+
+// make a scope the current scope. Replacement for `makeMain(hub)`
+export function setCurrentScope(scope: Scope): void;
+```
+
+The following APIs already exist but will behave differently:
+
+* `withScope()` will still work, but it will actually fork an execution context. So this will roughly do the same as doing `hub.run()` today in languages that have that, which forks an execution context. 
+
+APIs that are currently on the hub should instead be called directly on the scope (e.g. `scope.captureException()` etc.), or via a global method (e.g. `Sentry.captureException()`).
+
+## Clients
+
+Instead of a client being optional, there will now _always_ be a client. It may be a Noop Client that does nothing, if `init()` has not been called yet.
+
+A scope has a reference to a client. By default it will reference a noop client. You can bind a client to a scope via `scope.setClient()`. 
+The client is inherited by forked scopes.
```

Comment:
How does this behave if the client is replaced on the parent of a forked scope. Does the client on the forked scope inherit the new client or is the one from back when the fork happened still in use?

Reply:
I'd say that generally a client is always inherited, if a user overwrites it somewhere down the line than going forward this overwritten client will continue to be inherited. Basically:

```js
scope1 - client1
FORK -> scope2 - client1
  scope2.setClient(client2)
  FORK -> scope3 - client2
```

does that make sense?

Reply:
What I meant was:

```
scope1 - client1
FORK -> scope 2 - client 1
scope1.setClient(client2)
  scope2.getClient() <-- what does this return?
```

Reply:
sorry, wanted to show this but my note was not specific enough xD

```js
scope1.getClient() // >> client1
FORK -> scope2
  scope2.setClient(client2)
  scope2.getClient() // >> client2
  FORK -> scope3
    scope3.getClient() // >> client2
```

Reply:
Hmm but if you now set a different `client` on `scope1`, would that affect forked scopes (`scope2` and `scope3`) as well?

Reply:
ah, you mean would it affect _already forked_ scopes? I'd say not - I think forking should only happen at fork-time and after that the scopes should not affect each other anymore. Does that sound reasonable?

Reply:
It sure does sound easier to implement and reason about. Probably best then to start out this way as I think that's also how hubs behave at the moment.

Reply:
yes, they do, I think! 👍 I will update the RFC to make this explicit/clear.

Reply:
added a note describing this!